server/src/main/java/io/druid/server/router/ConsistentHashAvaticaConnectionBalancer.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.server.router;
+
+import io.druid.client.selector.Server;
+import io.druid.java.util.common.StringUtils;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+public class ConsistentHashAvaticaConnectionBalancer implements AvaticaConnectionBalancer
+{
+  private final ConsistentHasher hasher = new ConsistentHasher(null);
+
+  @Override
+  public Server balance(Collection<Server> servers, String connectionId)
+  {
+    synchronized (hasher) {
+      if (servers.isEmpty()) {
+        return null;
+      }
+      Map<String, Server> serverMap = new HashMap<>();
+      for (Server server : servers) {
+        serverMap.put(server.getHost(), server);
+      }
+
+      hasher.updateKeys(serverMap.keySet());
```

Comment:
this does not make sense to be called on each request ... broker appear/disappear very rarely and changes could be handled via the notifications.
also, then, i believe consistent hasher would be higher performing overall because broker-set doesn't change often and finding server only involves a hash and a binary search.

Reply:
hm, I'm wary of potential bugs with managing state with missed/dropped notifications, etc.

I feel like this is fine as-is for this PR and we could enhance this later (the docs do mention that the consistent hasher is experimental), benchmark shows that this updateKeys() call isn't particularly expensive for the use case, the number of brokers in a druid cluster isn't typically very large, with 10 brokers it's adding about <1us overhead per request:

```
  public Set<String> updateKeysBenchmark(Set<String> currentKeys)
  {
    Set<String> added = new HashSet<>(currentKeys);
    added.removeAll(previousKeys);

    Set<String> removed = new HashSet<>(previousKeys);
    removed.removeAll(currentKeys);

    for (String key : added) {
      addKey(key);
    }

    for (String key : removed) {
      removeKey(key);
    }

    // store a copy in case the input was immutable
    return previousKeys = new HashSet<>(currentKeys);
  }


    @Benchmark
  @BenchmarkMode(Mode.AverageTime)
  @OutputTimeUnit(TimeUnit.MICROSECONDS)
  public void updateKeys(Blackhole blackhole) throws Exception
  {
    for (int i = 0; i < numIds; i++) {
      blackhole.consume(hasher.updateKeysBenchmark(servers));
    }
  }

Benchmark                             (numIds)  Mode  Cnt      Score      Error  Units
ConsistentHasherBenchmark.updateKeys    100000  avgt   30  89350.198 ± 2640.216  us/o

```
> also, then, i believe consistent hasher would be higher performing overall because broker-set doesn't change often and finding server only involves a hash and a binary search.

```
Benchmark                       (numIds)  Mode  Cnt       Score      Error  Units
RendezvousHasherBenchmark.hash    100000  avgt   30  330105.546 ± 7494.327  us/op
ConsistentHasherBenchmark.hash    100000  avgt   30  43539.036 ± 1134.461  us/op
```

The consistentHasher does make faster hashing choices in the steady-state, but neither seem to impose much overhead. 

When running the RendezvousHasher benchmark on my laptop, I was getting ~3us per hashing request with 10 brokers, so I would expect the load balancing overhead to be dwarfed by the actual query processing times.

I made Rendezvous the default since it's a simpler implementation with less state to manage and less locking.

Reply:
if it have no performance impact then overhead is alright. thanks for benchmark.